lib/cc/yaml/nodes/mapping.rb

```diff
@@ -178,11 +175,37 @@ def each_scalar(type = nil, &block)
 
       protected
 
-        def dup_values
-          duped_mapping = @mapping.map { |key, value| [key.dup, value.dup] }
-          @mapping      = Hash[duped_mapping]
-          self
+      def dup_values
+        duped_mapping = @mapping.map { |key, value| [key.dup, value.dup] }
+        @mapping      = Hash[duped_mapping]
+        self
+      end
+
+      def check_duplicates(key)
+        warning("has multiple %p entries, keeping last entry", key) if self[key]
+      end
+
+      def check_incompatibility(key)
+        if creates_incompatibility?(key)
+          warning(incompatibility_message(key), key)
         end
+      end
+
+      def incompatibility_message(key)
+        "#{extant_engines_or_languages_key} key already found, dropping key: #{key}. " + INCOMPATIBLE_KEYS_WARNING
```

Comment:
It would be a tiny bit faster to format this with interpolation:

```ruby
"#{extant_engines_or_languages_key} key already found, dropping key: #{key}. #{INCOMPATIBLE_KEYS_WARNING}"
```